File: XMLCheck/Handler/CheckRulesXML01/Check_MA_THE_BHYT.cs
using System.Text.RegularExpressions;
using XMLCheck.Handler.CheckRulesXML1;
using XMLCheck.Models;
using XMLCheck.Services;

namespace XMLCheck.Handler.CheckRulesXML01
{
    public class Check_MA_THE_BHYT : Xml1RuleBase
    {
        public override string Key => "MA_THE_BHYT";

        public override ErrorDetails? Check(Xml1Model model, HoSoContext context)
        {
            if (string.IsNullOrWhiteSpace(model.MA_THE_BHYT))
            {
                return Error($"{Key} không được để trống");
            }

            // 1. Tách nhiều thẻ bằng ;
            var cards = model.MA_THE_BHYT
                .Split(';', StringSplitOptions.RemoveEmptyEntries)
                .Select(x => x.Trim())
                .ToList();

            // Rule: phải dùng dấu ;
            if (cards.Count > 1 && !model.MA_THE_BHYT.Contains(";"))
            {
                return Error($"Các {Key} phải cách nhau bằng dấu ';'");
            }

            foreach (var card in cards)
            {
                // 2. Độ dài tối thiểu
                if (card.Length < 15)
                {
                    return Error($"{Key} không đúng định dạng: {card}");
                }

                // 3. Ký tự 1–2: mã đối tượng BHYT
                var doiTuong = card.Substring(0, 2);
                if (!DictionaryCache.DoiTuongBHYT.ContainsKey(doiTuong))
                {
                    return Error($"MA_THE_BHYT có mã đối tượng không hợp lệ: {doiTuong}");
                }

                // 4. Ký tự 3: quyền lợi (1–5)
                var quyenLoi = card[2];
                if (quyenLoi < '1' || quyenLoi > '5')
                {
                    return Error("Ký tự thứ 3 của MA_THE_BHYT phải là số từ 1 đến 5");
                }

                // 6. 10 ký tự cuối phải là số
                var soCuoi = card.Substring(card.Length - 10, 10);
                if (!Regex.IsMatch(soCuoi, @"^\d{10}$"))
                {
                    return Error("10 ký tự cuối của MA_THE_BHYT phải là số");
                }
            }

            return null;
        }
    }
}

File: XMLValidator_v2/src/parsers/rules/xml1/Check_MA_THE_BHYT.ts
import { Xml1RuleBase } from './xml1RuleBase';
import { Xml1Model } from '../../../interfaces/xml-models.interface';
import { HoSoContext } from '../../../services/hoso-context';
import { ErrorDetails } from '../../../dto/xml-check.dto';
import { DictionaryCacheInstance } from '../../dictionary.cache';

export class Check_MA_THE_BHYT extends Xml1RuleBase {
  public get key(): string {
    return 'MA_THE_BHYT';
  }

  public check(model: Xml1Model, context: HoSoContext): ErrorDetails | null {
    if (!model.MA_THE_BHYT || !model.MA_THE_BHYT.trim()) {
      return this.error(`${this.key} không được để trống`);
    }

    const cards = model.MA_THE_BHYT
      .split(';')
      .map(x => x.trim())
      .filter(x => x.length > 0);

    if (cards.length > 1 && !model.MA_THE_BHYT.includes(';')) {
      return this.error(`Các ${this.key} phải cách nhau bằng dấu ';'`);
    }

    for (const card of cards) {
      if (card.length < 15) {
        return this.error(`${this.key} không đúng định dạng: ${card}`);
      }

      const doiTuong = card.substring(0, 2);
      if (!(doiTuong in DictionaryCacheInstance.doiTuongBHYT)) {
        return this.error(`MA_THE_BHYT có mã đối tượng không hợp lệ: ${doiTuong}`);
      }

      const quyenLoi = card.charAt(2);
      if (quyenLoi < '1' || quyenLoi > '5') {
        return this.error("Ký tự thứ 3 của MA_THE_BHYT phải là số từ 1 đến 5");
      }

      const soCuoi = card.substring(card.length - 10);
      if (!/^\d{10}$/.test(soCuoi)) {
        return this.error("10 ký tự cuối của MA_THE_BHYT phải là số");
      }
    }

    return null;
  }
}
